lib/THC/THCReduce.cuh

```diff
@@ -38,30 +79,22 @@ kernelReduceNoncontigDim(TensorInfo<T, IndexType> out,
                          T init,
                          ModifyOp modifyOp,
                          ReduceOp reduceOp) {
-  const IndexType sliceIndex = getReduceNoncontigDimSliceIndex<IndexType>();
+  IndexType sliceIndex  = blockIdx.x * blockDim.x + threadIdx.x;
+  IndexType sliceStride = gridDim.x * blockDim.x;
 
-  if (sliceIndex >= totalSlices) {
-    return;
-  }
+  for(;sliceIndex<totalSlices; sliceIndex+=sliceStride){
+    IndexType outOffset = IndexToOffset<T, IndexType, ADims>::get(sliceIndex, out);
+    IndexType inOffset = IndexToOffset<T, IndexType, BDims>::get(sliceIndex, in);
+    T r = init;
 
-  // Each thread picks a point in `out` and `in` for which it is
-  // producing the reduction
-  const IndexType outOffset =
-    IndexToOffset<T, IndexType, ADims>::get(sliceIndex, out);
-  const IndexType inBaseOffset =
-    IndexToOffset<T, IndexType, BDims>::get(sliceIndex, in);
-
-  // For each point in reductionSize, reduce into `r`
-  IndexType inOffset = inBaseOffset;
-  T r = init;
+    for(IndexType i = 0; i<reductionSize; i++){
+      r = reduceOp(r, modifyOp(in.data[inOffset]) );
+      inOffset += reductionStride;
+    }
 
-  for (IndexType i = 0; i < reductionSize; ++i) {
-    r = reduceOp(r, modifyOp(in.data[inOffset]));
-    inOffset += reductionStride;
+    __syncthreads();
```

Comment:
It is unfortunately necessary. We're trying to prevent warps from getting too far ahead which will have negative effects on the memory system.